synapse/event_auth.py

```diff
@@ -299,9 +301,12 @@ def _is_membership_change_allowed(
             raise AuthError(403, "%s is banned from the room" % (target_user_id,))
         return
 
-    if Membership.JOIN != membership:
+    # Require the user to be in the room for membership changes other than join/knock.
+    if Membership.JOIN != membership and Membership.KNOCK != membership:
```

Comment:
Does the various code in this function need to take into account the room version?